runtime/block_reduction.cu

```diff
@@ -48,8 +48,58 @@ __device__ void blockReduce(
       index_utils::maskedOffset<!X_REDUCE, !Y_REDUCE, !Z_REDUCE>(
           threadIdx, blockDim);
 
+  // number of reductions per block
+  unsigned int reduction_num =
+      index_utils::maskedSize<!X_REDUCE, !Y_REDUCE, !Z_REDUCE>(blockDim);
+
+  // smem_offset is the offset into shared memory for the current thread.
+  // To ensure coalesced access to shared memory, we need to ensure
+  // each transaction is accessing a contiguous block of 128 bytes.
+  // For outer reduction where TIDy is in the reduction dimension and TIDx
+  // is in the iteration dimension and TIDz is not used. We have
+  // reduction_tid = TIDy and reduction_idx = TIDx. If we directly use the
+  // offset based on reduction_tid and reduction_idx, we will have stride
+  // access to shared memory. For example:
+  // offset = reduction_idx * reduction_size + reduction_tid
+  //        = TIDx * blockDim.y + TIDy
+  // To avoid this, we should always use the offset based on the indexing of
+  // threads within a block.
   // Offset into smem for the current thread
-  unsigned int smem_offset = reduction_idx * reduction_size + reduction_tid;
+  unsigned int smem_offset = threadIdx.x + threadIdx.y * blockDim.x +
+      threadIdx.z * blockDim.x * blockDim.y;
+
+  // The peer stride is the distance between the current element and its
+  // reduction peer, it depends on the reduction dimension.
+  constexpr int num_redu_dims = (int)X_REDUCE + (int)Y_REDUCE + (int)Z_REDUCE;
+  // reduction in 3 dimensions, XYZ, stride is 1
+  unsigned int peer_stride = 1;
+  if (num_redu_dims == 1) {
+    // Reduction only in 1 dimension, X or Y or Z
+    // e.g. inner or outer reduction
+    // If X_REDUCE, reducing in neighbor cols in smem, peer_stride is 1
+    // If Y_REDUCE, reducing in neighbor rows in smem, peer_stride is blockDim.x
+    // If Z_REDUCE, reducing in neighbor planes in smem, peer_stride is
+    // blockDim.x * blockDim.y
+    peer_stride = X_REDUCE ? 1
+        : Y_REDUCE         ? blockDim.x
+                           : blockDim.x * blockDim.y;
+  } else if (num_redu_dims == 2) {
+    // Reduction in 2 dimensions, only one dimension is not reduced, !X, !Y, !Z
+    // If !Z_REDUCE, merge XY, reducing neighbor cols, peer_stride is 1
+    // If !X_REDUCE, merge ZY, reducing neighbor rows, peer_stride is blockDim.x
+    // If !Y_REDUCE, if blockDim.y == 1, merge XZ, peer_stride is 1, otherwise
+    //               different strides for different reduction stages, change
+    //               smem layout based on reduction_idx and reduction_tid, may
+    //               have stride access and bank conflict, rare case, may happen
+    //               for batch norm doing multiple reductions per block.
```

Comment:
Not quite sure what's intended here. Why changing the layout?

Reply:
added a figure for reference:
![image](https://github.com/NVIDIA/Fuser/assets/116412316/34f8572c-618e-4c5c-9883-9056ab4015cd)